policyeval/broker.go

```diff
@@ -0,0 +1,347 @@
+	// Eval needs to be Ack'd before this timer finishes.
+	nackTimer := time.AfterFunc(b.nackTimeout, func() {
+		b.Nack(eval.ID, token)
+	})
+
+	// Mark eval as not Ack'd yet.
+	b.unack[eval.ID] = &unackEval{
+		Eval:      eval,
+		Token:     token,
+		NackTimer: nackTimer,
+	}
+
+	// Increment dequeue counter.
+	b.enqueuedEvals[eval.ID] += 1
+
+	logger.Debug("eval dequeued",
+		"eval_id", eval.ID, "policy_id", eval.Policy.ID, "token", token)
+	return eval, token, nil
+}
+
+// findWork returns an eval from the queue heap or nil if there's no eval available.
+func (b *Broker) findWork(queue string) *sdk.ScalingEvaluation {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	pending, ok := b.pendingEvals[queue]
+	if !ok {
+		return nil
+	}
+
+	if pending.Len() == 0 {
+		return nil
+	}
+
+	// Pop heap and update reference.
+	raw := heap.Pop(&pending)
+	b.pendingEvals[queue] = pending
+
+	return raw.(*sdk.ScalingEvaluation)
+}
+
+// waitForWork blocks until queue receives an item or the context is canceled.
+func (b *Broker) waitForWork(ctx context.Context, queue string) bool {
+	b.logger.Debug("waiting for eval", "queue", queue)
+
+	b.l.Lock()
+	waitCh, ok := b.waiting[queue]
+	if !ok {
+		waitCh = make(chan struct{}, 1)
+		b.waiting[queue] = waitCh
+	}
+	b.l.Unlock()
+
+	select {
+	case <-ctx.Done():
+		return false
+	case <-waitCh:
+		return true
+	}
+}
+
+// Ack is used to acknowledge the successful completion of an eval.
+func (b *Broker) Ack(evalID, token string) error {
+	logger := b.logger.With("eval_id", evalID, "token", token)
+
+	logger.Debug("ack eval", "eval_id", evalID, "token", token)
+
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	// Lookup the unack'd eval.
+	unack, ok := b.unack[evalID]
+	if !ok {
+		return fmt.Errorf("evaluation ID not found")
+	}
+	if unack.Token != token {
+		return fmt.Errorf("token does not match for evaluation ID")
+	}
+
+	// Ensure we were able to stop the timer.
+	if !unack.NackTimer.Stop() {
+		return fmt.Errorf("evaluation ID Ack'd after Nack timer expiration")
+	}
+
+	// Cleanup.
+	delete(b.unack, evalID)
+	delete(b.enqueuedEvals, evalID)
+	delete(b.enqueuedPolicies, unack.Eval.Policy.ID)
+
+	b.logger.Debug("eval ack'd", "policy_id", unack.Eval.Policy.ID)
+	return nil
+}
+
+// Nack is used to mark an eval as not completed.
+func (b *Broker) Nack(evalID, token string) error {
+	logger := b.logger.With("eval_id", evalID, "token", token)
+
+	logger.Debug("nack eval")
+
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	// Lookup the unack'd eval
+	unack, ok := b.unack[evalID]
+	if !ok {
+		return fmt.Errorf("evaluation ID not found")
+	}
+	if unack.Token != token {
+		return fmt.Errorf("token does not match for evaluation ID")
+	}
+
+	logger = logger.With("policy_id", unack.Eval.Policy.ID)
+
+	// Stop the timer, doesn't matter if we've missed it.
+	unack.NackTimer.Stop()
+
+	// Cleanup.
+	delete(b.unack, evalID)
+
+	// Check if we've hit the delivery limit.
+	if dequeues := b.enqueuedEvals[evalID]; dequeues >= b.deliveryLimit {
```

Comment:
should the eval be GC'd from `enqueuedEvals` and `enqueuedPolicies`?

Reply:
Yup, good catch.